web/main.js

```diff
@@ -116,33 +124,58 @@ let send = async f => {
 let receive = e => {
 	if (!receiving) {
 		receiving = JSON.parse(new TextDecoder('utf8').decode(e.data));
-		receiving.data = new Uint8Array(receiving.size);
+		// Maybe there is a better way to identify a single file transfer.
+		receiving.id = document.getElementById("magiccode").value + '-' + Date.now().toString(16);
 		receiving.offset = 0;
 		receiving.li = document.createElement('li');
 		receiving.li.appendChild(document.createTextNode(`↓ ${receiving.name}`));
 		receiving.li.appendChild(document.createElement('progress'));
 		receiving.progress = receiving.li.getElementsByTagName("progress")[0];
 		document.getElementById("transfers").appendChild(receiving.li);
+
+		downloadServiceWorker.postMessage({
+			id: receiving.id,
+			type: 'metadata',
+			name: receiving.name,
+			size: receiving.size
+		});
+
+		// `<a download=...>` doesn't work with service workers on Chrome yet.
+		// See https://bugs.chromium.org/p/chromium/issues/detail?id=468227
+		// let a = document.createElement('a');
+		// a.href = `/download/${receiving.id}`;
+		// a.download = receiving.name;
+		// a.style.display = 'none';
+		// document.body.appendChild(a);
+		// a.click();
+		// document.body.removeChild(a);
+
+		// Possible solutions:
+
+		// - `window.open` is blocked as a popup.
+		// window.open(`/download/${receiving.id}`);
+
+		// - And this is quite scary but `Content-Disposition` to the rescue!
+		//   It will navigate to 404 page if there is no service worker for some reason...
+		//   But if `postMessage` didn't throw we should be safe.
+		window.location = `/download/${receiving.id}`;
```

Comment:
Content-Disposition sounds good to me. (#30 might be related to this but I haven't had time to look into it yet.)